File: Python/Map-2/first_char.py

# given an array of non-empty strings, return a Map<String, String> with a key for every different first character seen
# with the value of all the strings starting with that character appended together in the order they appear in the array
def first_char(str):
    d = {}
    
    for x in str:
        if not x[0] in d:
            d[x[0]] = x
        else:
            str_appended = d[x[0]] + x
            d[x[0]] = str_appended
    
    return d

print(first_char(['salt', 'tea', 'soda', 'toast']))
print(first_char(['aa', 'bb', 'cc', 'aAA', 'cCC', 'd']))
print(first_char([]))

File: JavaScript/Map-2/firstChar.js

// given an array of non-empty strings
// return a Map<String, String> with a key for every different first character seen
// with the value of all the strings starting with that character appended together in the order they appear in the array
const firstChar = (strings) => {
    const myMap = new Map();

    for (let i = 0; i < strings.length; i++) {
        if (!myMap.has(strings[i][0])) {
            myMap.set(strings[i][0], strings[i]);
        } else {
            const prevStr = myMap.get(strings[i][0]);

            myMap.set(strings[i][0], prevStr + strings[i]);
        }
    }

    return myMap;
};

console.log(firstChar(['salt', 'tea', 'soda', 'toast']));
console.log(firstChar(['aa', 'bb', 'cc', 'aAA', 'cCC', 'd']));
console.log(firstChar([]));
